### ISSUE TITLE
[v2] Sending `RNN.appLaunched` with no listeners registered.

### ISSUE THREAD

**Issue by kasperkronborg** [wix/react-native-navigation#1397]
### Issue Description

When building a basic sample project for iOS, following the project in the `playground` folder, the app throws the following warning `Sending 'RNN.appLaunched' with no listeners registered.`

### Steps to Reproduce / Code Snippets / Screenshots

* `git clone git@github.com:kasperkronborg/test-navigation.git`
* `yarn install`
* `react-native run-ios`

In the github project I have added some log statements to the iOS Library, that gives the indication of a race condition. It seems that `RNN.appLaunched` are called to early, when there has still not been any listeners attached through `Navigation.events().onAppLaunched()`.

---
### Environment
* React Native Navigation version: 2.0.54
* React Native version: 0.45.1
* Platform(s) (iOS, Android, or both?): iOS
* Device info (Simulator/Device? OS version? Debug/Release?): Simulator, iOS 10, Debug

**Comment by DanielZlotin** [comment]
I'm getting `Permission denied (publickey). fatal: Could not read from remote repository.`
Any chance you test it with RN44 ? maybe that was introduced by some change in RN..

**Comment by DanielZlotin** [comment]
I tested and it reproduced in RN44. Weird, as I can't reproduce it in the e2e playground app in the navigation..

**Comment by DanielZlotin** [comment]
Can't reproduce, 
But discovered a bug in RN45 causing cmd+R to crash in navigation. Will open a separate issue.

**Comment by casperstorm** [comment]
There is indeed something fishy going on.
I downloaded, and ran the test-project as mentioned above. The project launched at first, but I do as well get `Sending `RNN.appLaunched` with no listeners registered.`. I then closed the simulator, packager and ran the project again. This time the project did not get pass the launch screen.

It does indeed smell of some kind of race condition.

**Comment by DanielZlotin** [comment]
To summarize the discussion in discord, it does not reproduce in RN44.

**Comment by steffenmllr** [comment]
I also ran into this issue using `RNN 2.0.1950` and `RN 0.46.4` with a simple init script:

```
import Navigation from 'react-native-navigation';
import SampleScreen from './app/screens/sampleScreen';
Navigation.registerContainer('SampleScreen', () => SampleScreen);

Navigation.events().onAppLaunched(() => {
    Navigation.setRoot({
        container: {
            name: 'SampleScreen'
        }
    });
});
```

@DanielZlotin  what is the best way to deal with this? going back to `RN 0.45.X` or `RN 0.44.X` ?

**Comment by Paul-Todd** [comment]
Did anyone get any further with this - it still is a problem on 0.46.4 and 2.0.1977

**Comment by joonhocho** [comment]
@Paul-Todd I can confirm that having remote debugging on causes:
2017-09-09 01:56:39.618 [warn][tid:main][RCTEventEmitter.m:54] Sending `RNN.appLaunched` with no listeners registered.
and thus `Navigation.events().onAppLaunched()` is not being called.

**Comment by joonhocho** [comment]
@DanielZlotin Do you have any thoughts on this? I am trying out v2, and I can't use remote debugging because of this issue.

**Comment by joonhocho** [comment]
I am digging into it, and It seems `RCTJavaScriptDidLoadNotification` is called for each javascript bundle load, and enabling remote debugging or hmr causes separate js bundles to load other than the app bundle.
Enabling remote debugging causes its js bundle to load before app bundle (which makes sense cause you want to attach debugger before starting app) so that jsload event is raised before app successfully starts listening to applaunch event.
I think there must be a better way to start an app other than listening to `RCTJavaScriptDidLoadNotification` event because it can be false positive due the reasons above.
Or at least should be able to differentiate app bundle's `RCTJavaScriptDidLoadNotification` from that of other js bundles.

**Comment by joonhocho** [comment]
Created PR https://github.com/wix/react-native-navigation/pull/1816

**Comment by casperstorm** [comment]
On Jul 21, 2018 13:49, "stale[bot]" <notifications@github.com> wrote:

Closed #1397 <https://github.com/wix/react-native-navigation/issues/1397>.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/wix/react-native-navigation/issues/1397#event-1746057473>,
or mute the thread

**Comment by flexaddicted** [comment]
Hi @joonhocho! 'm working on a brownfield app and I have the same problem. I would like to now if you can point me to the code you've used in the react-native-navigation project in order to fix the problem. I've downloaded the source code but I cannot find anymore that fix.

Right now I'm using `RCTJavaScriptDidLoadNotification` in order to know if the bundle has been loaded or not but I receive the errore described in this thread.

Thanks, Lorenzo

**Comment by JoseFMP** [comment]
Any updates on this issue?
